### PR TITLE
add `whenPaused` modifier

### PR DESCRIPTION
**Pull request by Gornutz**
# Description
Add the `whenPaused` modifer to admin setters to ensure no mismatch of pools and stable asset.

## Type of change

- [X] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [X] The diff is legible and has no extraneous changes
- [X] Complex code has been commented, including external interfaces
- [X] Tests are included for all code paths
- [X] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

**Comment by 0xSpraggins** [comment]
I think it would be a better approach to remove the `setStableAsset` function all together and set `stableAsset` within the `setUniswapV3Pool` function adding an extra parameter with that stable asset and cross referencing to verify that the pool accurately contains the proper token. This way we don't need the `whenPaused` modifier, can reduce the number of calls the admin needs to make and guarantee that the Uniswap pool and our `stableAsset` match.

File: src/BlueberryStaking.sol [hunk 1]
```diff
@@ -747,24 +731,39 @@ contract BlueberryStaking is
...
         if (_observationPeriod == 0 || _observationPeriod > 432_000) {
             revert InvalidObservationTime();
         }
+        
+        address token0 = IUniswapV3Pool(_uniswapPool).token0();
+        address token1 = IUniswapV3Pool(_uniswapPool).token1();
+
+        if (_stableAsset != token0 && _stableAsset != token1) {
+            revert InvalidStableAsset();
+        }
```
Comment (0xSpraggins) [inline src/BlueberryStaking.sol:754]:
```suggestion
bool blbIsToken0 = IUniswapV3Pool(_uniswapPool).token0() == address(blb)

If (blbIsToken0) {
    address token1 = IUniswapV3Pool(_uniswapPool).token1();
    If (token1 != _stableAsset) revert InvalidStableAsset();
    stableAsset = token1
 } else {
    address token0 = IUniswapV3Pool(_uniswapPool).token0();
    If (token0 != _stableAsset) revert InvalidStableAsset();
    stableAsset = token0
}
```

We can optimize this future will less external calls and comparisons and then reuse blbIsToken0 when setting `uniswapV3Info`.

File: src/BlueberryStaking.sol [hunk 2]
```diff
@@ -747,24 +731,39 @@ contract BlueberryStaking is
...
 
         uniswapV3Info = UniswapV3PoolInfo({
             pool: _uniswapPool,
             observationPeriod: _observationPeriod,
             blbIsToken0: IUniswapV3Pool(_uniswapPool).token0() == address(blb)
         });
+
+        stableAsset = IERC20(_stableAsset);
```
Comment (0xSpraggins) [inline src/BlueberryStaking.sol:762]:
```suggestion
```
Can be removed if we make the optimization above

File: src/interfaces/IBlueberryStaking.sol [hunk 3]
```diff
@@ -87,8 +90,13 @@ interface IBlueberryStaking {
...
     event TreasuryUpdated(address treasury);
 
-    /// @notice Emitted when the admin updates the stable asset.
-    event StableAssetUpdated(address asset, uint256 decimals);
+    /// @notice Emitted when the admin updates the uniswap pool.
+    event UniswapV3PoolUpdated(
+        address _uniswapPool,
+        address _stableAsset,
+        uint8 decimals,
+        uint256 _observationPeriod
```
Comment (0xSpraggins) [inline src/interfaces/IBlueberryStaking.sol:98]:
```suggestion
        address uniswapPool,
        address stableAsset,
        uint8 decimals,
        uint256 observationPeriod
```

Small format fix

File: test/Control.t.sol [hunk 4]
```diff
@@ -183,18 +190,19 @@ contract Control is Test {
...
     }
 
-    // Test changing the stable token address
-    function testChangeStable() public {
+    // Test changing the uniswap pool
+    function testChangeUniswapPool() public {
         vm.startPrank(owner);
-
-        // Deploy a new stable coin 
-        MockToken token = new MockToken(9);
```
Comment (0xSpraggins) [inline test/Control.t.sol:196]: Would be good to also add a tests that expects to revert when setting a stable token that is not in the Uniswap pair

**Review by 0xSpraggins — CHANGES_REQUESTED** [review]

File: src/BlueberryStaking.sol [hunk 5]
```diff
@@ -747,24 +731,43 @@ contract BlueberryStaking is
...
+        } else {
+            address token0 = IUniswapV3Pool(_uniswapPool).token0();
+            if (token0 != _stableAsset) revert InvalidStableAsset();
+            stableAsset = IERC20(token0);
+        }
 
         uniswapV3Info = UniswapV3PoolInfo({
             pool: _uniswapPool,
             observationPeriod: _observationPeriod,
             blbIsToken0: IUniswapV3Pool(_uniswapPool).token0() == address(blb)
```
Comment (0xSpraggins) [inline src/BlueberryStaking.sol:764]:
```suggestion
            blbIsToken0: blbIsToken0
```
U can reuse the variable from line 749

File: test/Control.t.sol [hunk 6]
```diff
@@ -183,18 +192,28 @@ contract Control is Test {
...
+            USDC,
+            3600
+        );
 
         // Check that the stable token address was updated correctly
-        assertEq(address(blueberryStaking.stableAsset()), address(token));
+        assertEq(address(blueberryStaking.stableAsset()), USDC);
+
+        // Expect the pool to revert since the stable asset is not in that pool
+        vm.expectRevert();
```
Comment (0xSpraggins) [inline test/Control.t.sol:210]:
```suggestion
        vm.expectRevert(IBlueberryStaking.InvalidStableAsset.selector);
```

**Review by 0xSpraggins — CHANGES_REQUESTED** [review]

**Review by 0xSpraggins — APPROVED** [review]

**Review by IAm0x52 — APPROVED** [review]
Fix looks good. setStableAsset has been removed. setUniswapV3Pool now sets stableAsset as well as pool to prevent any mismatch.

**Pull request merged**